### PR TITLE
explicit which layer should model callables provide to attacks

### LINKED ISSUE #160: Be more clear when attacks need logits or softmax probs
Some attacks operate with the model's logits and others with the softmax probs. We should improve the documentation to make sure it is clear when the `model` argument of an attack class needs to be logits and when it needs to be softmax probs (see #157). In addition, we could use #154 to have some more automated checks in the `__init__` method of attacks that test whether the `model` argument is appropriate. There could also be some more generic checks looking for the TensorFlow softmax op.

### PR DESCRIPTION
**Pull request by npapernot**
Closes #160

### CODE REVIEW

File: cleverhans/attacks.py [hunk 1]
```diff
@@ -226,6 +236,9 @@ class FastGradientMethod(Attack):
     def __init__(self, model, back='tf', sess=None):
         """
         Create a FastGradientMethod instance.
```
Comment (goodfeli) [inline cleverhans/attacks.py]: It's more straightforward to just say `cleverhans.model.Model`.

Rather than documenting how the callable should behave, maybe we should just leave that out. If people want to duck-type the `Model` class, its call method returns probabilities, so they should do the right thing when reading its doc.

In places were we require probabilities, we can usually check that the output comes from some specific op, usually Softmax. Explicit runtime checks are a lot more reliable than documentation.

File: cleverhans/attacks_tf.py [hunk 2]
```diff
@@ -459,17 +462,16 @@ def __init__(self, sess, model, batch_size, confidence,
...
-                  targeted attack.
-        :param nb_classes: The number of classes the model has.
+        :param sess: a TF session.
+        :param model: a CleverHans model object.
+        :param batch_size: Number of attacks to run simultaneously.
         :param confidence: Confidence of adversarial examples: higher produces
                            examples with larger l2 distortion, but more
                            strongly classified as adversarial.
-        :param batch_size: Number of attacks to run simultaneously.
+        :param targeted: boolean controling the behavior of the adversarial
```
Comment (goodfeli) [inline cleverhans/attacks_tf.py]: controlling

File: cleverhans/attacks_tf.py [hunk 3]
```diff
@@ -459,17 +462,16 @@ def __init__(self, sess, model, batch_size, confidence,
...
 
-        :param x: (required) A tensor with the inputs.
-        :param y: (optional) A tensor with the true labels for an untargeted
-                  attack. If None (and y_target is None) then use the
-                  original labels the classifier assigns.
-        :param y_target: (optional) A tensor with the target labels for a
-                  targeted attack.
-        :param nb_classes: The number of classes the model has.
+        :param sess: a TF session.
+        :param model: a CleverHans model object.
```
Comment (goodfeli) [inline cleverhans/attacks_tf.py]: Maybe "instance of cleverhans.model.Model" to be maximally straightforward

**Comment by npapernot** [comment]
All comments have been addressed

**Comment by goodfeli** [comment]
I think the build just stalled, so pushing a dummy commit should fix it. LGTM, OK to merge once the tests pass.
